File: backend/src/algorithms/domains.ts
import type { Grid, Slot } from "../types/puzzle.js";

function matchesPrefilledLetters(grid: Grid, slot: Slot, word: string): boolean {
  return slot.cells.every(({ row, column }, index) => {
    const cell = grid[row]?.[column] ?? "";
    return cell === "" || cell === word[index];
  });
}

export function buildInitialDomains(
  grid: Grid,
  slots: Slot[],
  words: string[],
): Map<string, string[]> {
  return new Map(
    slots.map((slot) => [
      slot.id,
      words.filter(
        (word) =>
          word.length === slot.length &&
          matchesPrefilledLetters(grid, slot, word),
      ),
    ]),
  );
}

File: backend/app/algorithms/domains.py
from typing import List, Dict
from app.models import Slot

def build_initial_domains(
    grid: List[List[str]],
    slots: List[Slot],
    words: List[str],
) -> Dict[str, List[str]]:
    domains = {}
    for slot in slots:
        candidates = []
        for word in words:
            if len(word) != slot.length:
                continue
            
            # Ensure candidate word matches any pre-filled cells in the grid
            matches = True
            for idx, cell in enumerate(slot.cells):
                prefilled = grid[cell.row][cell.column]
                if prefilled != "" and prefilled != "#" and word[idx] != prefilled:
                    matches = False
                    break
            if matches:
                candidates.append(word)
        domains[slot.id] = candidates
    return domains
